lib/src/command/add.dart

```diff
@@ -0,0 +1,237 @@
+    await runner.run(['get']);
+  }
+
+  /// Creates a new in-memory [Pubspec] by adding [package] to [original].
+  Future<Pubspec> _addPackagesToPubspec(
+      Pubspec original, PackageInfo package) async {
+    ArgumentError.checkNotNull(original, 'original');
+    ArgumentError.checkNotNull(package, 'package');
+
+    final dependencies = [...original.dependencies.values];
+    var devDependencies = [...original.devDependencies.values];
+
+    if (isDevelopment) {
+      final dependencyNames = dependencies.map((dependency) => dependency.name);
+
+      if (dependencyNames.contains(package.name)) {
+        log.message('${package.name} is already in dependencies. '
+            'Please remove existing entry before adding it to dev_dependencies');
+        await flushThenExit(exit_codes.SUCCESS);
+      }
+
+      devDependencies.add(package.toPackageRange(cache));
+    } else {
+      final devDependencyNames =
+          devDependencies.map((devDependency) => devDependency.name);
+
+      if (devDependencyNames.contains(package.name)) {
+        log.message('${package.name} was found in dev_dependencies. '
+            'Removing ${package.name} and adding it to dependencies instead.');
+        devDependencies =
+            devDependencies.takeWhile((d) => d.name != package.name).toList();
+      }
+
+      dependencies.add(package.toPackageRange(cache));
+    }
+
+    return Pubspec(
+      original.name,
+      version: original.version,
+      sdkConstraints: original.sdkConstraints,
+      dependencies: dependencies,
+      devDependencies: devDependencies,
+      dependencyOverrides: original.dependencyOverrides.values,
+    );
+  }
+
+  /// Parse [PackageInfo] from [package].
+  ///
+  /// [package] must be written in the format
+  /// `<package-name>[:<version-constraint>]`, where quotations should be used
+  /// if necessary.
+  ///
+  /// Examples:
+  /// ```bash
+  /// retry
+  /// retry:2.0.0
+  /// retry:^2.0.0
+  /// retry:'>=2.0.0'
+  /// retry:'>2.0.0 <3.0.1'
+  /// 'retry:>2.0.0 <3.0.1'
+  /// retry:any
+  /// ```
+  ///
+  /// If a version constraint is provided when the `--path` or any of the
+  /// `--git-<option>` options are used, a [PackageParseError] will be thrown.
+  ///
+  /// If both `--path` and any of the `--git-<option>` options are defined,
+  /// a [PackageParseError] will be thrown.
+  ///
+  /// If any of the other git options are defined when `--git-url` is not
+  /// defined, an error will be thrown.
+  PackageInfo _parsePackage(String package) {
+    ArgumentError.checkNotNull(package, 'packages');
+
+    var git = <String, String>{};
+    if (gitUrl != null) git['url'] = gitUrl;
+    if (gitRef != null) git['ref'] = gitRef;
+    if (gitPath != null) git['path'] = gitPath;
+    if (git.isEmpty) git = null;
+
+    Map<String, String> hostInfo;
+    if (hostUrl != null) hostInfo = {'url': hostUrl};
+
+    PackageInfo parsedPackage;
+
+    try {
+      parsedPackage = PackageInfo.from(package,
```

Comment:
It seems like you're transforming command line options into a map (`git`) and then parsing it all in a function.

I think moving all this into this file is fine.. and not creating an intermediate map structure is probably preferable too.. Otherwise, I would convert it to a typed structure using an non-exported class.